### PR TITLE
[FLINK-26284] Introduces mark-for-deletion state in ZooKeeperStateHandleStore

### PR DESCRIPTION
**Pull request by XComp**
## What is the purpose of the change

The `ZooKeeperStateHandleStore` cleanup is not repeatable: The ZK node is deleted removing the `StateHandle` metadata before discarding the `StateHandle` itself. If the deletion of the `StateHandle` fails, there is no way to determine the location for a subsequent cleanup anymore.

## Brief change log

We fixed this by introducing another child node collecting all the lock child nodes. A state node now has the following structure:
* `/<name>` - root node referencing the actual `StateHandle`
  * `/<name>/locks` - The locks child which contains all the lock nodes. This child node is `PERSISTENT` like the parent node to prevent the node going away automatically after all connections are cleared. We want to do the cleanup explicitly.
    * `/<name>/locks/<lock-child>` - A node referencing to a single ZK connection. This child lock is going to be `EPHEREMAL` like in the previous version and will be cleaned up automatically as soon as the ZK client is disconnected.

Removing the `/<name>/locks` means that the corresponding node is ready for deletion. Marking the node as deletable is only possible if all lock nodes are removed.

The interface contract stays the same or is hardened (for the`replace` method).

Additionally, a minor fix for FLINK-26285 was added.

## Verifying this change

`ZooKeeperStateHandleStoreTest` was extended by multiple tests covering the execution on nodes that are marked for deletion: See the individual commits for the different test cases.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by XComp** [comment]
`ZooKeeperCompletedCheckpointStoreITCase.testSuspendKeepsCheckpoints` was missed when fixing ZK-related tests. This is covered now.

File: flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java
Comment (dmvk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java:420]: oh man, goto statement spotted 🙈

File: flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java
Comment (dmvk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java:418]: how did this ever work? Do we have a test case covering this?

File: flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java
Comment (dmvk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java:248]: Why do we need this check now?

File: flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java
Comment (dmvk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java:591]: Under which circumstances can this return null? (as non existent node seems to throw an exception)

File: flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java
Comment (dmvk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java:526]: It took me a while to understand what's meant as idempotent here. How about:

```suggestion
    private void deleteIfExists(String path) throws Exception {
```

?

File: flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java
Comment (dmvk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java:615]: Maybe `getRootLockPath` + `getInstanceLockPath` could be more explicit?

File: flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java
Comment (dmvk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java:477]:
```suggestion
        // We can now "commit" the removal by removing the root directory.
        deleteIdempotently(path);
```

File: flink-runtime/src/test/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStoreTest.java
Comment (dmvk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStoreTest.java:212]: nit: assertThrows?

File: flink-runtime/src/test/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStoreTest.java
Comment (dmvk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStoreTest.java:1217]: hmm this seems inconsistent with what I did in k8s implementation 🤔

so if I understand it correctly, we always want `getAllHandles` to return everything

**Review by dmvk — COMMENTED** [review]
Thanks @XComp for the PR, this already looks really good 👍 I've added some comments and questions, please take a look

File: flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java:420]: Yes, but I left it there because I didn't want to touch the code more than necessary considering that we might want to get rid of it, anyway, soon'ish

Comment (dmvk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java:420]: 👍 sure, this was an off-topic comment

File: flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java:526]: I like that. It's probably more descriptive than the `deleteIdempotently` method name 👍

File: flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java:591]: This was meant for hardening the implementation. In the current setup it cannot happen, because this method is only called by `replace` which is only used in `DefaultJobGraphStore` when the corresponding `JobGraph` is already added to `DefaultJobGraphStore.addedJobGraphs` (which means that the JobGraph was already locked before through `getAndLock` in `DefaultJobGraphStore.recoverJobGraph` or `addAndLock` in `DefaultJobGraphStore.putJobGraph`)

Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java:591]: This was meant for hardening the implementation. In the current setup it cannot happen, because this method is only called by `replace` which is only used in `DefaultJobGraphStore` when the corresponding `JobGraph` is already added to `DefaultJobGraphStore.addedJobGraphs` (which means that the JobGraph was already locked before through `getAndLock` in `DefaultJobGraphStore.recoverJobGraph` or `addAndLock` in `DefaultJobGraphStore.putJobGraph`)

File: flink-runtime/src/test/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStoreTest.java
Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStoreTest.java:212]: That's not possible here, is it? Because we need to verify the cause tree. But the one underneath should work... 👍

File: flink-runtime/src/test/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStoreTest.java
Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStoreTest.java:1217]: Good point. That's the shortcoming of the `ZooKeeperStateHandleStore` not collecting the handles in a transaction. But we concluded that it's not necessary to exclude the deleted once because it can always happen getting the handles and before utilizing them that the underlying handle was deleted.

I mentioned that already in the `ZooKeeperStateHandleStore` implementation but extended the `StateHandleStore` JavaDoc accordingly. With the interface JavaDoc changes, it should be ok to have this being implemented more strictly in the k8s implementation

File: flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java:248]: That was added after investigating the code paths. Initially, I wanted to make the `replace` call fail if the caller doesn't have a lock on that node. This wouldn't happen because of the `get(path, false)` call below. I could have replaced it by `getAndLock` which would set the lock if the caller didn't set it, yet, or fail if the node is marked for deletion.

But that would have changed the semantics of the `replace` method slightly. Instead, I went for hardening the implementation by adding this preconditions. See [my comment](https://github.com/apache/flink/pull/18869/files#r813931306) on the `hasLock` change thread below to see why the `replace` method is always called after a lock is already acquired.

File: flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java:178]: it's only called during the checkpoint creation (where we shouldn't run into the situation that a checkpoint gets deleted because only one leader access a checkpoint at a time) and when a JobGraph is added during a job submission.

So, in this case, this is only added to make the implementation consistent.

File: flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java:408]: Till's comment on that:
> I think the problem before was that when a JM lost its leadership then we only suspended it w/o releasing all resources. Now, the JM will shut down once it loses leadership. If it regains the leadership, then it will be restarted and read the state from ZooKeeper again. Due to this, I think the locking mechanism is probably no longer needed (assuming that no two leader sessions can overlap).

So that's kind of unrelated to this bug. The `getAllAndLock` is only used in the context of recovering checkpoints. This only happens during a failover in which case checkpoints are not deleted. The process that lost leadership would just release the checkpoints.

There is another code branch where (old) checkpoints are deleted when adding a new checkpoint. I haven't found the check, but I assume, only the leader is able to create checkpoints. Hence, there shouldn't be a clash.

But the fix is still needed to make it consistent.

Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java:408]: I added a test to cover this use which includes a timeout which I'm not that happy about. I'm thinking of just dropping it? 🤔

**Comment by XComp** [comment]
@dmvk thanks for looking at the PR. I addressed or responded to your comments. PTAL

File: flink-runtime/src/test/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStoreTest.java
Comment (dmvk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStoreTest.java:1278]: Sounds good, this seems to be only used by the `SessionDispatcherLeaderProcess` and the `StateHandleStore` itself

**Review by dmvk — APPROVED** [review]
Thanks for the update, I have few more minor comments, but LGTM overall. Good job 👍

File: flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java]: Fair enough - alternatively, I could have added a `@Nullable` annotation to the parameter. But I went for your approach 👍

File: flink-runtime/src/test/java/org/apache/flink/runtime/persistence/TestingLongStateHandleHelper.java
Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/persistence/TestingLongStateHandleHelper.java]: Good idea.

**Comment by XComp** [comment]
Thanks for the approval, @dmvk . I addressed your comments (and also updated the `StateHandleStore.getAllHandles` JavaDoc as discussed off-topic).

**Comment by XComp** [comment]
I reorganized the commits once more moving the FLINK-26285 fix into its own PR #18919 and rebased this branch onto the FLINK-26285 PR (that's why there's an additional commit still in the diff of this PR here). PR #18919 should be merged to `master` before merging this branch, probably, to be consistent in the process.

File: flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java:410]: This was addressed by refactoring the `getAllAndLock` method and addressing the change in PR #18919

**Comment by XComp** [comment]
I did another minor refactoring in order to be consistent with the changes of #18901 (see [comment](https://github.com/apache/flink/pull/18901/files#r814672839)): `addAndLock` is going to throw an `AlreadyExistException` in any case. I updated the commit message accordingly:
```
About ZooKeeperStateHandleStore.addAndLock:
  There shouldn't be a case where there is a concurrent delete on an
  addAndLock call because of the following reasons:
  - called through DefaultJobGraphStore.putJobGraph: Happens when a new
    job is submitted, i.e. a new JobID is used.
  - called through CheckpointCoordinator.addCompletedCheckpointToStoreAndSubsumeOldest:
    Each leader creates a new checkpoint ID using the CheckpointIDCounter,
    i.e. another JM cleaning up wouldn't have the handle to the newly
    created checkpoint.
  - called through CheckpointCoordinator.restoreSavepoint:
    A Savepoint is never subject to deletion
  That's why we keep the AlreadyExistException as a safe-guard
```

**Comment by XComp** [comment]
> I did another minor refactoring in order to be consistent with the changes of https://github.com/apache/flink/pull/18901 (see [comment](https://github.com/apache/flink/pull/18901/files#r814672839)): addAndLock is going to throw an AlreadyExistException in any case. I updated the commit message accordingly:

This was reverted after a follow-up discussion with @dmvk

**Comment by XComp** [comment]
rebased branch after PR #18919 was merged

**Comment by XComp** [comment]
Test failure are unrelated to this change:
* FLINK-26393
* [FLINK-26321](https://issues.apache.org/jira/browse/FLINK-26321)

**Pull request merged**
